src/goto-harness/memory_snapshot_harness_generator.cpp

```diff
@@ -62,114 +64,154 @@ void memory_snapshot_harness_generatort::handle_option(
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "harness function `" + id2string(harness_function_name) +
+      "` already in "
+      "the symbol table",
+      "--" GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT);
+  }
+}
+
+void memory_snapshot_harness_generatort::add_init_section(
+  goto_modelt &goto_model) const
+{
+  goto_functionst &goto_functions = goto_model.goto_functions;
+  symbol_tablet &symbol_table = goto_model.symbol_table;
+
+  goto_functiont &goto_function = goto_functions.function_map[entry_function_name];
+  const symbolt &function_symbol = symbol_table.lookup_ref(entry_function_name);
+
+  goto_programt &goto_program = goto_function.body;
+
+  // introduce a symbol for a Boolean variable to indicate the point at which
+  // the function initialisation is completed
+  symbolt &func_init_done_symbol = get_fresh_aux_symbol(
+    bool_typet(),
+    id2string(entry_function_name),
+    "func_init_done",
+    function_symbol.location,
+    function_symbol.mode,
+    symbol_table);
+  func_init_done_symbol.is_static_lifetime = true;
+  func_init_done_symbol.value = false_exprt();
 
-  auto ins_it2 = goto_program.insert_after(ins_it1);
-  ins_it2->make_assignment(code_assignt(var, true_exprt()));
+  const symbol_exprt func_init_done_var = func_init_done_symbol.symbol_expr();
 
-  auto ins_it3 = goto_program.insert_after(ins_it2);
-  ins_it3->make_goto(goto_program.const_cast_target(it));
+  // initialise func_init_done_var in __CPROVER_initialize if it is present
+  // so that it's FALSE value is visible before the harnessed function is called
+  goto_programt &cprover_initialize =
+    goto_functions.function_map.find(INITIALIZE_FUNCTION)->second.body;
+  cprover_initialize.insert_before(
+    std::prev(cprover_initialize.instructions.end()),
+    goto_programt::make_assignment(
+      code_assignt(func_init_done_var, false_exprt())));
+
+  const goto_programt::const_targett start_it =
+    goto_program.instructions.begin();
+
+  auto ins_it1 = goto_program.insert_before(
+    start_it,
+    goto_programt::make_goto(goto_program.const_cast_target(start_it)));
+  ins_it1->guard = func_init_done_var;
+
+  auto ins_it2 = goto_program.insert_after(
+    ins_it1,
+    goto_programt::make_assignment(
+      code_assignt(func_init_done_var, true_exprt())));
+
+  goto_program.insert_after(
+    ins_it2,
+    goto_programt::make_goto(goto_program.const_cast_target(
+      location_number.has_value() ? *goto_program.get_target(*location_number)
+                                  : start_it)));
 }
 
-void memory_snapshot_harness_generatort::add_assignments_to_globals(
-  const symbol_tablet &snapshot,
-  code_blockt &code) const
+code_blockt memory_snapshot_harness_generatort::add_assignments_to_globals(
+  const symbol_tablet &snapshot) const
 {
+  code_blockt code;
   for(const auto &pair : snapshot)
   {
     const symbolt &symbol = pair.second;
 
     if(!symbol.is_static_lifetime)
```

Comment:
If I understand it correctly, this is checking for a global right? I think this required one more property but I'm not sure which one it was (I think `is_lvalue` or something). I'm not sure if this matters here.

Reply:
seconded "is_lvalue" - I recollect this from Fotis & Hannes' recent Harness PR.